File: runtime/typescript/src/admission.ts
import type { InteractionEnvelope } from './types.js';
import { requestFingerprint } from './fingerprint.js';
import { IKError, type ProfileDescriptor, validateEnvelopeCore } from './validate.js';

export interface ContractValidationPort {
  validatePayload(profile: ProfileDescriptor, data: unknown): void | Promise<void>;
  validateArtifacts?(profile: ProfileDescriptor, envelope: InteractionEnvelope): void | Promise<void>;
}
export interface IdentityAuthenticator {
  authenticate(transportContext: Record<string, unknown>): string | Promise<string>;
}
export interface AuthorityAuthorizer {
  authorize(principal: string, envelope: InteractionEnvelope): void | Promise<void>;
}
export interface IdempotencyRecord { fingerprint: string; interactionId: string; receipt?: unknown; }
export interface IdempotencyStore {
  lookup(scope: string, key: string): IdempotencyRecord | undefined | Promise<IdempotencyRecord | undefined>;
  reserve(scope: string, key: string, record: IdempotencyRecord): void | Promise<void>;
}
export interface AdmissionResult { principal: string; fingerprint?: string; replay: boolean; prior?: IdempotencyRecord; }

export class AdmissionPipeline {
  constructor(
    private readonly contracts: ContractValidationPort,
    private readonly identity: IdentityAuthenticator,
    private readonly authority: AuthorityAuthorizer,
    private readonly idempotency: IdempotencyStore,
  ) {}

  async admit(envelope: InteractionEnvelope, profile: ProfileDescriptor, transportContext: Record<string, unknown>): Promise<AdmissionResult> {
    validateEnvelopeCore(envelope, profile);
    await this.contracts.validatePayload(profile, envelope.data ?? {});
    if (this.contracts.validateArtifacts) await this.contracts.validateArtifacts(profile, envelope);
    const principal=await this.identity.authenticate(transportContext);
    await this.authority.authorize(principal,envelope);
    const key=envelope.idempotency_key;
    if (!key) return {principal,replay:false};
    const fingerprint=await requestFingerprint(envelope);
    const target=envelope.target?.system ?? 'broadcast';
    const scope=`${target}:${envelope.profile.id}:${envelope.profile.version}`;
    const prior=await this.idempotency.lookup(scope,key);
    if (prior) {
      if (prior.fingerprint!==fingerprint) throw new IKError('IK_IDEMPOTENCY_CONFLICT','same idempotency key with divergent semantic request');
      return {principal,fingerprint,replay:true,prior};
    }
    await this.idempotency.reserve(scope,key,{fingerprint,interactionId:envelope.id});
    return {principal,fingerprint,replay:false};
  }
}

File: runtime/python/src/interaction_kernel/admission.py
from __future__ import annotations
from dataclasses import dataclass
from typing import Any, Mapping, Protocol
from .errors import IKError
from .fingerprint import request_fingerprint
from .validator import ContractValidator

class IdentityAuthenticator(Protocol):
    def authenticate(self, transport_context: Mapping[str, Any]) -> str: ...

class AuthorityAuthorizer(Protocol):
    def authorize(self, principal: str, envelope: Mapping[str, Any]) -> None: ...

class IdempotencyStore(Protocol):
    def lookup(self, *, scope: str, key: str) -> Mapping[str, Any] | None: ...
    def reserve(self, *, scope: str, key: str, fingerprint: str, interaction_id: str) -> None: ...

@dataclass(frozen=True)
class AdmissionResult:
    principal: str
    fingerprint: str | None
    replay: bool
    prior: Mapping[str, Any] | None = None

class AdmissionPipeline:
    def __init__(self, validator: ContractValidator, authenticator: IdentityAuthenticator, authorizer: AuthorityAuthorizer, idempotency: IdempotencyStore) -> None:
        self.validator = validator
        self.authenticator = authenticator
        self.authorizer = authorizer
        self.idempotency = idempotency

    def admit(self, envelope: Mapping[str, Any], transport_context: Mapping[str, Any]) -> AdmissionResult:
        self.validator.validate_envelope(envelope)
        principal = self.authenticator.authenticate(transport_context)
        self.authorizer.authorize(principal, envelope)
        key = envelope.get("idempotency_key")
        if not key:
            return AdmissionResult(principal, None, False)
        fingerprint = request_fingerprint(envelope)
        scope = f"{envelope['target']['system']}:{envelope['profile']['id']}:{envelope['profile']['version']}"
        prior = self.idempotency.lookup(scope=scope, key=key)
        if prior:
            if prior.get("fingerprint") != fingerprint:
                raise IKError("IK_IDEMPOTENCY_CONFLICT", "same idempotency key with divergent semantic request")
            return AdmissionResult(principal, fingerprint, True, prior)
        self.idempotency.reserve(scope=scope, key=key, fingerprint=fingerprint, interaction_id=envelope["id"])
        return AdmissionResult(principal, fingerprint, False)
